theano/compile/builders.py

```diff
@@ -124,17 +124,14 @@ def make_node(self, *inputs):
                          list(inputs) + self.shared_inputs,
                          [type() for type in self.output_types])
 
-    def make_thunk(self, node, storage_map, compute_map, no_recycling):
-        ret = super(OpFromGraph, self).make_thunk(node, storage_map,
-                                                  compute_map, no_recycling)
-        if not hasattr(self, "fn"):
-            self.fn = orig_function(self.new_inputs,
-                                    self.new_outputs,
-                                    **self.kwargs)
-        return ret
+    def prepare_node(self, node, storage_map, compute_map, impl):
+        if not hasattr(node.tag, "fn") and impl == 'py':
+            node.tag.fn = orig_function(self.new_inputs,
+                                        self.new_outputs,
+                                        **self.kwargs)
```

Comment:
This could go on self, because it doesn't depend on the inputs.  It would also avoid recompiling for multiple uses of it in the same graph.